# Deep Dive into Pandas

We'll get our first look at pandas' data structures.

Pandas is a Python library that provides extensive means for data analysis. Data scientists often work with data stored in table formats like .csv, .tsv, or .xlsx. Pandas makes it very convenient to load, process, and analyze such tabular data using SQL-like queries. In conjunction with Matplotlib and Seaborn,  Pandas provides a wide range of opportunities for visual analysis of tabular data.

The main data structures in Pandas are implemented with Series and DataFrame classes. The former is a one-dimensional indexed array of some fixed data type. The latter is a two-dimensional data structure - a table - where each column contains data of the same type. You can see it as a dictionary of Series instances. DataFrames are great for representing real data: rows correspond to instances (objects, observations, etc.), and columns correspond to features for each of the instances.

In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [4]:
%matplotlib inline

In [5]:
pd.options.display.max_rows = 10
pd.options.display.float_format = '{:,.2f}'.format #supress scientific notation
plt.rcParams['figure.figsize'] = (16, 12)

## Reading Data

Pandas has support for reading from many data sources, including

- `pd.read_csv`
- `pd.read_excel`
- `pd.read_html`
- `pd.read_json`
- `pd.read_hdf`
- `pd.read_sql`

For this section we'll work with some loan data. Dream Housing Finance company deals in all home loans. They have presence across all urban, semi urban and rural areas. Customer first apply for home loan after that company validates the customer eligibility for loan. 

In [6]:
loan = pd.read_csv("data/train.csv")
loan

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849               0.00         nan            360.00   
1               4583           1,508.00      128.00            360.00   
2               3000               0.00       66.00            360.00   
3               2583           2,358.00      120.00            360.00   
4               6000               0.00      141.00            360.00   
..               ...                ...         ...               ...   
609             2900               0.00       71.00            360.00   
610             4106               0.00       40.00            180.00   
611             8072             240.00      253.00            360.00   
612             7583               0.00      187.00            360.00   
613             4583               0.00      133.00            360.00   

     Credit_History Property_Area Loan_Status  
0              1.00         Urban           Y  
1              1.00         Rural           N  
2              1.00         Urban           Y  
3              1.00         Urban           Y  
4              1.00         Urban           Y  
..              ...           ...         ...  
609            1.00         Rural           Y  
610            1.00         Rural           Y  
611            1.00         Urban           Y  
612            1.00         Urban           Y  
613            0.00     Semiurban           N  

[614 rows x 13 columns]

## Data Structures

![A dataframe is made up of data, row labels, and column labels](figures/dataframe.png)

`read_csv` returned a `DataFrame`, which is somewhat similar to a spreadsheet or database table.
`pd.DataFrame` is the data container you'll work most with, and consists of a few components:

The data are in the middle of the table.
Each column of the data is a `pd.Series`, kind of like a 1-dimensional version of a DataFrame.

In [7]:
loan['LoanAmount']

0        nan
1     128.00
2      66.00
3     120.00
4     141.00
       ...  
609    71.00
610    40.00
611   253.00
612   187.00
613   133.00
Name: LoanAmount, Length: 614, dtype: float64

Both `pd.DataFrame`s and `pd.Series` have *row labels*, which can be accessed with the `.index` attribute:

In [8]:
loan.index

RangeIndex(start=0, stop=614, step=1)

`flights.index` is a `pd.Index` (there are many specialized index types, like `pd.RangeIndex`, but we'll talk about those later).

DataFrames store their column labels in a `.columns` attribute, which is also a `pd.Index`:

In [9]:
loan.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

## Data Types

![](figures/dtypes.png)

Like NumPy (but unlike regular Python lists), you'll want to know the `dtypes` of your data.
Improving on NumPy, pandas DataFrames can store *heterogenous* data;
each column of a DataFrame will have it's own type (int, float, datetime, bool, etc.), but the DataFrame can hold a mixture of these.

In [10]:
loan.shape

(614, 13)

In [11]:
loan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
Loan_ID              614 non-null object
Gender               601 non-null object
Married              611 non-null object
Dependents           599 non-null object
Education            614 non-null object
Self_Employed        582 non-null object
ApplicantIncome      614 non-null int64
CoapplicantIncome    614 non-null float64
LoanAmount           592 non-null float64
Loan_Amount_Term     600 non-null float64
Credit_History       564 non-null float64
Property_Area        614 non-null object
Loan_Status          614 non-null object
dtypes: float64(4), int64(1), object(8)
memory usage: 62.4+ KB


With this same method, we can easily see if there are any missing values. Here, Loan_ID ,Education,ApplicantIncome,CoapplicantIncome,Property_Area,Loan_Status there are none because each column contains 614 observations, the same number of rows we saw before with shape.

## Variable Identification
First, identify Predictor (Input) and Target (output) variables. Next, identify the data type and category of the variables.


In [12]:
loan.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849               0.00         nan            360.00   
1             4583           1,508.00      128.00            360.00   
2             3000               0.00       66.00            360.00   
3             2583           2,358.00      120.00            360.00   
4             6000               0.00      141.00            360.00   

   Credit_History Property_Area Loan_Status  
0            1.00         Urban           Y  
1            1.00         Rural           N  
2            1.00         Urban           Y  
3            1.00         Urban           Y  
4            1.00         Urban           Y

## Comaprison of pandas and oracle sql

In [13]:
loan.info() #describe loan;

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
Loan_ID              614 non-null object
Gender               601 non-null object
Married              611 non-null object
Dependents           599 non-null object
Education            614 non-null object
Self_Employed        582 non-null object
ApplicantIncome      614 non-null int64
CoapplicantIncome    614 non-null float64
LoanAmount           592 non-null float64
Loan_Amount_Term     600 non-null float64
Credit_History       564 non-null float64
Property_Area        614 non-null object
Loan_Status          614 non-null object
dtypes: float64(4), int64(1), object(8)
memory usage: 62.4+ KB


In [14]:
loan #select * from loan;(select all row and all columns from loan)

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849               0.00         nan            360.00   
1               4583           1,508.00      128.00            360.00   
2               3000               0.00       66.00            360.00   
3               2583           2,358.00      120.00            360.00   
4               6000               0.00      141.00            360.00   
..               ...                ...         ...               ...   
609             2900               0.00       71.00            360.00   
610             4106               0.00       40.00            180.00   
611             8072             240.00      253.00            360.00   
612             7583               0.00      187.00            360.00   
613             4583               0.00      133.00            360.00   

     Credit_History Property_Area Loan_Status  
0              1.00         Urban           Y  
1              1.00         Rural           N  
2              1.00         Urban           Y  
3              1.00         Urban           Y  
4              1.00         Urban           Y  
..              ...           ...         ...  
609            1.00         Rural           Y  
610            1.00         Rural           Y  
611            1.00         Urban           Y  
612            1.00         Urban           Y  
613            0.00     Semiurban           N  

[614 rows x 13 columns]

In [15]:
loan.head(5) #select * from data_filtered where WHERE rownum <= 5; ( select first 5 rows with all columns from loan)

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849               0.00         nan            360.00   
1             4583           1,508.00      128.00            360.00   
2             3000               0.00       66.00            360.00   
3             2583           2,358.00      120.00            360.00   
4             6000               0.00      141.00            360.00   

   Credit_History Property_Area Loan_Status  
0            1.00         Urban           Y  
1            1.00         Rural           N  
2            1.00         Urban           Y  
3            1.00         Urban           Y  
4            1.00         Urban           Y

In [16]:
loan[["Education","LoanAmount"]] #select Education,LoanAmount from loan; ( select a few columns from loan )

Education  LoanAmount
0        Graduate         nan
1        Graduate      128.00
2        Graduate       66.00
3    Not Graduate      120.00
4        Graduate      141.00
..            ...         ...
609      Graduate       71.00
610      Graduate       40.00
611      Graduate      253.00
612      Graduate      187.00
613      Graduate      133.00

[614 rows x 2 columns]

In [17]:
loan[loan["LoanAmount"]==128] #select * from loan where LoanAmount = 128 (select all rows and all columns from loan where LoanAmount =128

Loan_ID Gender Married Dependents     Education Self_Employed  \
1    LP001003   Male     Yes          1      Graduate            No   
211  LP001711   Male     Yes         3+      Graduate            No   
224  LP001750   Male     Yes          0      Graduate            No   
259  LP001864   Male     Yes         3+  Not Graduate            No   
264  LP001872   Male      No          0      Graduate           Yes   
..        ...    ...     ...        ...           ...           ...   
383  LP002234   Male      No          0      Graduate           Yes   
461  LP002484   Male     Yes         3+      Graduate            No   
482  LP002537   Male     Yes          0      Graduate            No   
543  LP002755   Male     Yes          1  Not Graduate            No   
602  LP002953   Male     Yes         3+      Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1               4583           1,508.00      128.00            360.00   
211             3430           1,250.00      128.00            360.00   
224             6250               0.00      128.00            360.00   
259             4931               0.00      128.00            360.00   
264             5166               0.00      128.00            360.00   
..               ...                ...         ...               ...   
383             7167               0.00      128.00            360.00   
461             7740               0.00      128.00            180.00   
482             2083           3,150.00      128.00            360.00   
543             2239           2,524.00      128.00            360.00   
602             5703               0.00      128.00            360.00   

     Credit_History Property_Area Loan_Status  
1              1.00         Rural           N  
211            0.00     Semiurban           N  
224            1.00     Semiurban           Y  
259             nan     Semiurban           N  
264            1.00     Semiurban           Y  
..              ...           ...         ...  
383            1.00         Urban           Y  
461            1.00         Urban           Y  
482            1.00     Semiurban           Y  
543            1.00         Urban           Y  
602            1.00         Urban           Y  

[11 rows x 13 columns]

In [18]:
loan[(loan["LoanAmount"]==128) & (loan["Education"]=="Graduate")][["Loan_ID","Married"]] #select Loan_ID,Married from loan where LoanAmount = 128 and Education='Graduate' ( select Loan_ID, Married of Graduates who LoanAmount = 128 )

Loan_ID Married
1    LP001003     Yes
211  LP001711     Yes
224  LP001750     Yes
264  LP001872      No
380  LP002226     Yes
383  LP002234      No
461  LP002484     Yes
482  LP002537     Yes
602  LP002953     Yes

In [19]:
loan.groupby("Education").size() #select LoanAmount, count(*) from loan group by Education; ( count customers according to Education)

Education
Graduate        480
Not Graduate    134
dtype: int64


### Grouping

In general, grouping data in Pandas goes as follows:



```python
df.groupby(by=grouping_columns)[columns_to_show].function()
```


1. First, the `groupby` method divides the `grouping_columns` by their values. They become a new index in the resulting dataframe.
2. Then, columns of interest are selected (`columns_to_show`). If `columns_to_show` is not included, all non groupby clauses will be included.
3. Finally, one or several functions are applied to the obtained groups per selected columns.

Here is an example where we group the data according to the values of the `Churn` variable and display statistics of three columns in each group:

In [20]:
columns_to_show = ['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount']

loan.groupby(['Loan_Status'])[columns_to_show].describe(percentiles=[])

ApplicantIncome                                              \
                      count     mean      std    min      50%       max   
Loan_Status                                                               
N                    192.00 5,446.08 6,819.56 150.00 3,833.50 81,000.00   
Y                    422.00 5,384.07 5,765.44 210.00 3,812.50 63,337.00   

            CoapplicantIncome                                            \
                        count     mean      std  min      50%       max   
Loan_Status                                                               
N                      192.00 1,877.81 4,384.06 0.00   268.00 41,667.00   
Y                      422.00 1,504.52 1,924.75 0.00 1,239.50 20,000.00   

            LoanAmount                                   
                 count   mean   std   min    50%    max  
Loan_Status                                              
N               181.00 151.22 85.86  9.00 129.00 570.00  
Y               411.00 144.29 85.48 17.00 126.00 700.00

Let’s do the same thing, but slightly differently by passing a list of functions to `agg()`:

In [21]:
columns_to_show = ['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount']

loan.groupby(['Loan_Status'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

ApplicantIncome                      CoapplicantIncome           \
                       mean      std amin   amax              mean      std   
Loan_Status                                                                   
N                  5,446.08 6,819.56  150  81000          1,877.81 4,384.06   
Y                  5,384.07 5,765.44  210  63337          1,504.52 1,924.75   

                           LoanAmount                     
            amin      amax       mean   std  amin   amax  
Loan_Status                                               
N           0.00 41,667.00     151.22 85.86  9.00 570.00  
Y           0.00 20,000.00     144.29 85.48 17.00 700.00

## Pandas Cheat Sheet

![](figures/cheat-sheet-preview.png)